core/api_token_api.php

```diff
@@ -0,0 +1,162 @@
+<?php
+# MantisBT - A PHP based bugtracking system
+
+# MantisBT is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# MantisBT is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with MantisBT.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * API Token API
+ *
+ * @package CoreAPI
+ * @subpackage ApiTokenAPI
+ * @copyright Copyright 2000 - 2002  Kenzaburo Ito - kenito@300baud.org
+ * @copyright Copyright 2002  MantisBT Team - mantisbt-dev@lists.sourceforge.net
+ * @link http://www.mantisbt.org
+ *
+ * @uses crypto_api.php
+ */
+
+require_api( 'crypto_api.php' );
+
+define( 'API_TOKEN_LENGTH', 32 );
+
+/**
+ * Create an API token
+ *
+ * @param string $p_token_name The name (description) identifying what the token is going to be used for.
+ * @param integer $p_user_id The user id.
+ * @return string The plain token.
+ * @access public
+ */
+function api_token_create( $p_token_name, $p_user_id ) {
+	if( is_blank( $p_token_name ) ) {
+		error_parameters( lang_get( 'token_name' ) );
+		trigger_error( ERROR_EMPTY_FIELD, ERROR );
+	}
+
+	$t_token_name = trim( $p_token_name );
+
+	$t_plain_token = crypto_generate_uri_safe_nonce( API_TOKEN_LENGTH );
+	$t_hash = api_token_hash( $t_plain_token );
+	$t_date_created = db_now();
+
+	$t_query = 'INSERT INTO {api_token}
+					( user_id, name, hash, date_created )
+					VALUES ( ' . db_param() . ', ' . db_param() . ', ' . db_param() . ', ' . db_param() . ' )';
+	db_query( $t_query, array( $p_user_id, (string)$t_token_name, $t_hash, $t_date_created ) );
+
+	return $t_plain_token;
+}
+
+/**
+ * Calculate the hash for the token.
+ * @param string $p_token The plain token.
+ * @return string The one way hash for the token
+ * @access public
+ */
+function api_token_hash( $p_token ) {
+	# We ignore spaces in tokens.  We inject spaces for users for readability only.
+	return sha1( str_replace( ' ', '', $p_token ) );
```

Comment:
If we want to make sure that an attacker does not obtain the plain-text value of the token, we should not relay on SHA-1. Maybe use the `hash` function with a `sha256` argument?